packages/sdk/src/libs/clients/lsp/client.ts

```diff
@@ -0,0 +1,103 @@
+import { LongShortPair__factory, LongShortPair } from "@uma/core/contract-types/ethers";
+import type { SignerOrProvider, GetEventType } from "../..";
+import { Event } from "ethers";
+import { Balances } from "../../utils";
+
+export type Instance = LongShortPair;
+
+export type TokensCreated = GetEventType<Instance, "TokensCreated">;
+export type TokensRedeemed = GetEventType<Instance, "TokensRedeemed">;
+export type ContractExpired = GetEventType<Instance, "ContractExpired">;
+export type PositionSettled = GetEventType<Instance, "PositionSettled">;
+
+export function connect(address: string, provider: SignerOrProvider): Instance {
+  return LongShortPair__factory.connect(address, provider);
+}
+
+export interface EventState {
+  sponsors?: string[];
+  longs?: Balances;
+  shorts?: Balances;
+  collateral?: Balances;
+  expired?: boolean;
+  expiredBy?: string;
+}
+
+export function reduceEvents(state: EventState = {}, event: Event, index?: number): EventState {
+  switch (event.event) {
+    case "TokensCreated": {
+      const typedEvent = event as TokensCreated;
+      const { sponsor, collateralUsed, tokensMinted } = typedEvent.args;
+
+      const sponsors = new Set(state.sponsors || []);
+      const longs = Balances(state.longs || {});
+      const shorts = Balances(state.shorts || {});
+      const collateral = Balances(state.collateral || {});
+
+      sponsors.add(sponsor);
+      longs.add(sponsor, tokensMinted);
+      shorts.add(sponsor, tokensMinted);
+      collateral.add(sponsor, collateralUsed);
+
+      return {
+        ...state,
+        collateral: collateral.balances,
+        shorts: shorts.balances,
+        longs: longs.balances,
+        sponsors: Array.from(sponsors.values()),
+      };
+    }
+    case "TokensRedeemed": {
+      const typedEvent = event as TokensRedeemed;
```

Comment:
what do you mean

Reply:
i just mean it would be nice to have typescript know the set of different event types that are allowed to be passed in so we don't have to cast. But since this isn't the _entire_ set of events that are possible, we'd need to do some runtime validation to guarantee that. I think the way you're doing it now makes sense, just thinking out loud :)